__tests__/webhooks.ts

```diff
@@ -0,0 +1,127 @@
+  Source,
+  User,
+  UserMarketingCta,
+} from '../src/entity';
+import { sourcesFixture } from './fixture/source';
+import { usersFixture } from './fixture/user';
+import { DataSource } from 'typeorm';
+import createOrGetConnection from '../src/db';
+import request from 'supertest';
+import { createHmac } from 'crypto';
+
+let app: FastifyInstance;
+let con: DataSource;
+
+beforeAll(async () => {
+  con = await createOrGetConnection();
+  app = await appFunc();
+  return app.ready();
+});
+
+afterAll(() => app.close());
+
+beforeEach(async () => {
+  jest.resetAllMocks();
+  await saveFixtures(con, Source, sourcesFixture);
+  await saveFixtures(con, User, usersFixture);
+  await saveFixtures(con, MarketingCta, [
+    {
+      campaignId: 'worlds-best-campaign',
+      variant: 'card',
+      status: MarketingCtaStatus.Active,
+      createdAt: new Date('2024-03-13 12:00:00'),
+      flags: {
+        title: 'Join the best community in the world',
+        description: 'Join the best community in the world',
+        ctaUrl: 'http://localhost:5002',
+        ctaText: 'Join now',
+      },
+    },
+  ]);
+  await saveFixtures(
+    con,
+    UserMarketingCta,
+    usersFixture.map((user) => ({
+      marketingCtaId: 'worlds-best-campaign',
+      userId: user.id,
+      createdAt: new Date('2024-03-13 12:00:00'),
+    })),
+  );
+});
+
+describe('POST /webhooks/customerio/marketing_cta', () => {
+  describe('webhook signature verification', () => {
+    it('should return 403 when no x-cio-timestamp header', async () => {
+      const { body } = await request(app.server)
+        .post('/webhooks/customerio/marketing_cta')
+        .set('x-cio-signature', '123')
+        .send({
+          userId: '1',
+          marketingCtaId: 'worlds-best-campaign',
+        })
+        .expect(403);
+
+      expect(body.error).toEqual('Invalid signature');
+    });
+
+    it('should return 403 when no x-cio-signature header', async () => {
+      const { body } = await request(app.server)
+        .post('/webhooks/customerio/marketing_cta')
+        .set('x-cio-timestamp', '123')
+        .send({
+          userId: '1',
+          marketingCtaId: 'worlds-best-campaign',
+        })
+        .expect(403);
+
+      expect(body.error).toEqual('Invalid signature');
+    });
+
+    it('should return 200 when signature is valid (static)', async () => {
+      const timestamp = 1715784949;
+      const payload = {
+        userId: '1',
+        marketingCtaId: 'worlds-best-campaign',
+      };
+      const hash =
+        'b01e7dfc3ce67eaab1010b7021b86ce2ba12677d191e0e04b14cd859e347f866';
+      const { body } = await request(app.server)
+        .post('/webhooks/customerio/marketing_cta')
+        .set('x-cio-timestamp', timestamp.toString())
+        .set('x-cio-signature', hash)
+        .send(payload)
+        .expect(200);
+
+      expect(body.success).toEqual(true);
+    });
+
+    it('should return 200 when signature is valid (dynamic)', async () => {
```

Comment:
what does `dynamic` mean in test name?

Reply:
The other one (static) has the values hardcoded. And this one uses real timestamp for hmac signature

Reply:
So its the same test?

Reply:
Basically yeah

Reply:
Never enough tests but interested what is the benefit of having both? Any edge case that you found?

Reply:
No edge cases, but my reasoning is if we modify the signature checker we'll also modify the signature generator in the test, so we might introduce the same error in both places. Then the static test will stand as a safeguard

Reply:
Makes sense 👍